counter: add generic implementation and use it to replace StringCounter

tklauser opened the pull request:
The generic implementation will be used in a follow-up change to replace `StringCounter` instances tracking backends in `pkg/service` to use `counter.Counter[netip.Addr]`.

Keep IntCounter as is for now because for `IntCounter.ToBPFData` it requires the map elements to be ordered (i.e. `constraints.Ordered`), not just comparable.  This method is only used on instances of type `IntCounter` so far and making the map elements ordered would limit its use e.g. with `netip.Addr`.

aspsk commented on pkg/counter/counter.go:
```diff
@@ -0,0 +1,44 @@
+type Counter[T comparable] map[T]int
+
+// Add increments the reference count for the specified key.
+func (c Counter[T]) Add(key T) (changed bool) {
```
Won't it be more readable like this?
```
func (c Counter[T]) Add(key T) bool {
        value, exists := c[key]
        c[key] = value + 1
        return !exists
}
```
(for me `changed` sounds like the value was changed)

tklauser replied:
Agree, thanks for the suggestion.

aspsk reviewed (approved):
Looks good, one comment about code, but your's variant looks fine as well

tklauser committed:
counter: add generic implementation and use it to replace StringCounter

The generic implementation will be used in a follow-up change to replace
StringCounter instances tracking backends in pkg/service  to use
counter.Counter[netip.Addr].

Keep IntCounter as is for now because for IntCounter.ToBPFData it
requires the map elements to be ordered (i.e. constraints.Ordered), not
just comparable.  This method is only used on instances of type
IntCounter so far and making the map elements ordered would limit its
use e.g. with netip.Addr.

Signed-off-by: Tobias Klauser <tobias@isovalent.com>

ldelossa reviewed (approved):
Nice!
